---
title: "MCF10A Breast Carcinogenome Portal"
output:
  html_document:
  toc: true
  theme: united
---

An interactive portal for retrieving and visualizing data from mammary cell carcinogenome project

### Dataset details

This experiment uses 345 selected chemicals for breast carcinogenicity testing, including 120 breast carcinogens, 114 non-carcinogens, and 68 miscellaneous chemicals (e.g. nuclear receptor ligands, BU SRP chemicals, lung carcinogens). Chemical carcinogenicity and genotoxicity annotations are based on the Carcinogenicity Potency Database ([CPDB](https://toxnet.nlm.nih.gov/cpdb/)), which is the result of tissue-specific long-term animal cancer tests in rodents, or breast carcinogens published from [Rudel et. al., 2007](https://onlinelibrary.wiley.com/doi/10.1002/cncr.22653/pdf). In the CRCGN project, MCF10A (breast epithelial) cells are exposed to each individual chemical for 24 hours and their gene expression is profiled on the L1000 platform. Each chemical is assayed at 3 doses (3 fold dilutions starting from the highest concentration of 100uM, with the exception of selected BUSRP chemicals, which had custom chosen starting concentrations) with triplicate profiles generated for each dose. This experiment is repeated also on MCF10A TP53-/- cells. For each chemical and dose profile, the gene expression of 1000 landmark genes is calculated as a moderated z-score (weighted collapsed z-score of the 3 replicate perturbational profiles with respect to the entire plate).

### Chemical Annotation

A table detailing the chemicals in the mammary cell carcinogenicity screening and their associated annotation.

### Differential Expression

A list of differentially expressed genes for a given chemical of interest.

### Gene set enrichment

Gene set enrichment scores for a given chemical of interest. Gene sets include the [MSigDB](http://software.broadinstitute.org/gsea/msigdb) collections (Hallmark, C2 reactome pathways), and gene targets of various nuclear receptors ([NURSA](https://www.nursa.org/nursa/transcriptomine/index.jsf)). Enrichment scores were computed based on multiple methods, including gsva, ssGSEA, zscore (from R Bioconductor package, [GSVA](https://bioconductor.org/packages/release/bioc/html/GSVA.html)), and gsproj (custom script).

### Heatmap (interactive)

A heatmap visualizer to explore gene set enrichment results. For details, see [Morpheus](https://software.broadinstitute.org/morpheus/).

### Heatmap (static)

A static heatmap visualizer to explore gene set enrichment results, with option to subset samples to several premade chemical sets.

### Connectivity (In progress)

Summarizes the similarity of L1000 mammary cell line profiles to profiles in the CMAP database (containing ~9000 unique annotated perturbagens, 84 perturbagen classes). Results limited to the top 50 (highest connectivity) and bottom 50 (lowest connectivity), representing the top hits for signature concordance and signature reversal respectively. 


---

Credits: Amy Li, Stefano Monti, David Sherr, Broad Institute CMap team.

Contact us at [ajli@bu.edu](mailto:ajli@bu.edu)

This project is supported by Superfund Research Program at Boston University ([BUSRP](http://www.bu.edu/sph/research/research-landing-page/superfund-research-program-at-boston-university)), [NIH/LINCS](http://www.lincsproject.org), and [Art beCAUSE Breast Cancer Foundation](https://artbecause.org)


